### ISSUE TITLE
Navigation approach

### ISSUE THREAD

**Issue by ForbesLindesay** [tmpvar/jsdom#1903]
I'm working on making navigation and form post work in jsdom, so I can upstream things instead of adding them to zombie.js.  You can see what I've done so far in #1902.

The next step is to implement the hook for loading pages that get navigated to, and including them in the history.  What this means is that the JSDOM object may ultimately get new window objects, as the navigation occurs.  I think this is the simplest approach.  My suggestion is that there should also be a hook `onNewWindow` that is called when a link is opened in a new tab.  This hook will be given a reference to a separate JSDOM instance.

What this means is that JSDOM changes from representing a single loaded window, to being equivalent to a single tab within a modern, tabbed browser.  I don't think this notion of history can be simply made external, while keeping jsdom's implementation of the history and hash-change APIs.

**Comment by domenic** [comment]
Amazing. This is exactly what we've dreamed of. Having the JSDOM object change what it represents makes perfect sense to me.

The ideal solution would be to implement the WindowProxy so that the window is === itself before and after navigation. I think that can be layered on top though, later.

Can you say more about what the onNewWindow hook would be useful for?

**Comment by ForbesLindesay** [comment]
If you imagine a implementing a full tabbed browser in JavaScript, by rendering a JSDOM dom to canvas or something: onNewWindow means that a new tab should be opened, displaying the contents of the new dom.  e.g. if you rendered the following page using jsdom:

```html
<a href="http://example.com" target="__blank">Clicking me opens a new tab</a>
```

**Comment by ForbesLindesay** [comment]
I've started looking at this, but the spec is really hard to follow because there doesn't seem to be a class in jsdom called "Browsing Context".  I think currently all that functionality is tangled up with the window object.  Navigating across pages will mean multiple windows per browsing context, so I think it needs to be a separate object.

Is the right approach to make this a class and put it in `lib/jsdom/browser/BrowsingContext`?  I can try and create it according to `https://html.spec.whatwg.org/#windows` but I'm not 100% clear what each step of each algorithm involves.

**Comment by domenic** [comment]
That sounds right, yeah. The interesting part will be hooking it up to iframes. Also note that browsing contexts are 1:1 with WindowProxy objects.

And yeah, the spec is known to be pretty confusing. Anne and I are working to fix it but it's a long-term project. I'm at breakfast right now but will be in #jsdom IRC later today if you want more real-time discussion.

**Comment by domenic** [comment]
The biggest advice I think will be working on writing tests for observable differences, either from the outside or as web platform tests.

**Comment by ForbesLindesay** [comment]
> Also note that browsing contexts are 1:1 with WindowProxy objects.

That's useful to know.  I guess in that case it's about separating out a clear distinction between WindowProxy and Window object.  I'm not sure that's made very clear yet.

**Comment by domenic** [comment]
Right, jsdom does not have any WindowProxy implementation at all right now, just Window. WindowProxy isn't too hard to implement, I think, at least in the basics. I am a little worried about the performance consequences though, so I'd want to hold off on implementing it until we had some larger-scale benchmarks in place mimicking our biggest consumers (e.g. ReSpec and Ecmarkup).
